test/pylib/pool.py

```diff
@@ -64,41 +93,73 @@ async def get(self, *args, **kwargs) -> T:
 
-    async def steal(self) -> None:
-        """Take ownership of a previously borrowed object.
-           Frees up space in the pool.
+    async def put(self, obj: T, is_dirty: bool):
+        """Return a previously borrowed object to the pool
+           if it's not dirty, otherwise destroy the object
+           and free up space in the pool.
         """
+        if is_dirty:
+            await self.destroy(obj)
+
         async with self.cond:
-            self.total -= 1
+            if is_dirty:
+                self.total -= 1
+            else:
+                self.pool.append(obj)
             self.cond.notify()
 
-    async def put(self, obj: T):
-        """Return a previously borrowed object to the pool."""
+    async def replace_dirty(self, obj: T, *args, **kwargs) -> T:
+        """Atomically `put` a previously borrowed dirty object and `get` another one.
+           The 'atomicity' guarantees that the space freed up by the returned object
+           is used to return another object to the caller. The caller doesn't need
+           to wait for space to be freed by another user of the pool.
+
+           Note: the returned object might have been constructed earlier or it might
+           be built right now, as in `get`.
+           *args and **kwargs are used as in `get`.
+        """
+        await self.destroy(obj)
+
         async with self.cond:
-            self.pool.append(obj)
-            self.cond.notify()
+            if self.pool:
+                self.total -= 1
+                return self.pool.pop()
+
+            # Need to construct a new object.
+            # The space for this object is already accounted for in self.total.
 
-    def instance(self, *args, **kwargs) -> AsyncContextManager[T]:
+        return await self._build_and_get(*args, **kwargs)
+
+    def instance(self, dirty_on_exception: bool, *args, **kwargs) -> AsyncContextManager[T]:
         class Instance:
-            def __init__(self, pool):
+            def __init__(self, pool: Pool[T], dirty_on_exception: bool):
                 self.pool = pool
+                self.dirty = False
+                self.dirty_on_exception = dirty_on_exception
 
             async def __aenter__(self):
                 self.obj = await self.pool.get(*args, **kwargs)
                 return self.obj
 
             async def __aexit__(self, exc_type, exc, obj):
                 if self.obj:
-                    await self.pool.put(self.obj)
+                    self.dirty |= self.dirty_on_exception and exc is not None
+                    await self.pool.put(self.obj, is_dirty=self.dirty)
                     self.obj = None
 
-        return Instance(self)
+        return Instance(self, dirty_on_exception)
+
+    async def _build_and_get(self, *args, **kwargs) -> T:
+        """Precondition: we allocated space for this object
+           (it's included in self.total).
+        """
+        try:
+            obj = await self.build(*args, **kwargs)
```

Comment:
I just realized that self.build should be self._build because it's not intended for public use - as I looked at your name _build_and_get